### PR TITLE
Add role ocp4-workload-mlops

### PR DESCRIPTION
**Pull request by tsailiming**
##### SUMMARY

First PR to include the MLOps Worlshop

##### ISSUE TYPE

- New role Pull Request

##### COMPONENT NAME

ocp4-workload-mlops  , for OCP 4.6 

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+# - name: import stack image
+#   shell: |
+#     oc import-image --all quarkus-stack -n openshift
+
+- name: Pre-create and warm user workspaces
+  include_tasks: create_che_workspace.yaml
+  vars:
+    user: "{{ item }}"
+  with_list: "{{ users }}"
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml:158]: Change from `with_list` to `loop`...

```
- name: Pre-create and warm user workspaces
  include_tasks: create_che_workspace.yaml
  loop: "{{ users }}"
  loop_control:
    loop_var: user

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+#     name: quarkus-stack
+#     namespace: openshift
+#   register: r_stack_is
+#   retries: 200
+#   delay: 10
+#   until: r_stack_is.resources | list | length == 1
+
+# - name: import stack image
+#   shell: |
+#     oc import-image --all quarkus-stack -n openshift
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  k8s:
+    state: present
+    merge_type:
+    - strategic-merge
+    - merge
+    definition: "{{ lookup('file', item ) | from_yaml }}"
+  loop:
+    - ./files/pipelines_subscription.yaml
+    - ./files/odh_subscription.yaml
+    - ./files/seldon_subscription.yaml
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml:12]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# - name: "Wait for ODH operator to be Succeeded"
+#   k8s_facts:
+#     api_version: operators.coreos.com/v1alpha1
+#     kind: ClusterServiceVersion
+#     name: "{{odh_operator_ver}}"
+#     namespace: openshift-operators
+  #register: result
+  #until: result | json_query("resources[0].status.phase=='Succeeded'")
+  #retries: 40
+  #delay: 10
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/install-guides.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    -e USER_DEPLOY_REPO_NAME={{user_deploy_repo_name}} \
+    -e ARGOCD_URL=https://argocd-server-labs-argocd.{{ route_subdomain }} \
+    -e NEXUS_URL=https://nexus-labs-nexus.{{ route_subdomain }} \
+    -e NEXUS_DOCKER_REGISTRY=nexus-docker-labs-nexus.{{ route_subdomain }} \
+    -e RHDM_URL=https://dm-rhdmcentr-labs-dm.{{ route_subdomain }} \
+    -e GIT_USER_PASSWORD={{ gogs_pwd }}
+
+ #   -e KEYCLOAK_URL=https://keycloak-labs-infra.{{ route_subdomain }} \    
+#   -e CHE_USER_NAME={{ workshop_che_user_name }}
+#   -e RHAMT_URL=http://rhamt-web-console-labs-infra.{{ route_subdomain }}
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-guides.yaml]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/pre_workload.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
+---
+# Implement your Pre Workload deployment tasks here
+
+# - name: Ensuring AWS region is defined
+#   fail:
+#     msg: "This workload requires aws_region to be defined. Exiting..."
+#   when: aws_region is not defined
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/pre_workload.yml]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/pre_workload.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  set_fact:
+    master_url: "{{ r_api_url.resources[0].status.apiServerURL }}"
+
+- name: set the console
+  set_fact:
+    console_url: "https://{{ r_console_route.resources[0].spec.host }}"
+
+# - name: set bastion_fqdn
+#   set_fact:
+#     bastion_fqdn: "{{ subdomain_base }}"
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/pre_workload.yml]: Removed.

File: ansible/roles/ocp4-workload-mlops/tasks/install-dm.yaml [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        kind: ImageStream
+        name: rhdm-kieserver-rhel8
+        namespace: labs-dm
+        state: absent
+
+    - name: Import rhdm-kieserver-rhel8
+      command: |
+        oc import-image rhdm-7/rhdm-kieserver-rhel8:{{rhdm_rhelserver_tag}} \ 
+        --from=registry.redhat.io/rhdm-7/rhdm-kieserver-rhel8:{{rhdm_rhelserver_tag}} \
+        --reference-policy=local --confirm -n labs-dm
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-dm.yaml:94]: Fixed

File: ansible/roles/ocp4-workload-mlops/tasks/install-guides.yaml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    -e WORKSHOPS_URLS='https://raw.githubusercontent.com/rh-mlops-workshop/ocp-al-ml-workshop-{{guide}}-guides/master/_ocp-ai-ml-workshop.yml' \
+    -e JUPYTERHUB_URL=https://jupyterhub-labs-infra.{{ route_subdomain }} \
+    -e GRAFANA_URL="https://grafana-route-labs-grafana.{{ route_subdomain }}//d/U1cSDzyZz/prediction-analytics" \
+    -e USER_MODEL_REPO_NAME={{user_source_repo_name}} \
+    -e USER_DEPLOY_REPO_NAME={{user_deploy_repo_name}} \
+    -e ARGOCD_URL=https://argocd-server-labs-argocd.{{ route_subdomain }} \
+    -e NEXUS_URL=https://nexus-labs-nexus.{{ route_subdomain }} \
+    -e NEXUS_DOCKER_REGISTRY=nexus-docker-labs-nexus.{{ route_subdomain }} \
+    -e RHDM_URL=https://dm-rhdmcentr-labs-dm.{{ route_subdomain }} \
+    -e GIT_USER_PASSWORD={{ gogs_pwd }}
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-guides.yaml:32]: Fixed

File: ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    - "user.info: Grafana url is https://grafana-route-labs-grafana.{{ route_subdomain }}"
+    - "user.info: Admin login is {{grafana_admin_user}}/{{grafana_admin_password}}"
+    - "user.info: "
+    - "user.info: OpenShift credentials for attendees: {{workshop_openshift_user_name}} / {{ workshop_openshift_user_password }}"
+    - "user.info: CodeReady Workspaces credentials for attendees: {{workshop_che_user_name}} / {{ workshop_che_user_password }}"
+    - "user.info: "
+    - "user.info: CodeReady Console: https://codeready-labs-infra.{{ route_subdomain }}"
+    - "user.info: Admin login with admin/admin"
+    - "user.info: "
+    - "user.info: Workshop guide and materials are available here: https://red.ht/mlopsworkshop"
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml:37]: Fixed

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+- name: Create operator subscription for CodeReady
+  k8s:
+    state: present
+    merge_type:
+    - strategic-merge
+    - merge
+    definition: "{{ lookup('file', item ) | from_yaml }}"
+  loop:
+  - ./files/codeready_operatorgroup.yaml
+  - ./files/codeready_subscription.yaml
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+# deploy codeready CR
+- name: Create CR for CodeReady
+  k8s:
+    state: present
+    merge_type:
+    - strategic-merge
+    - merge
+    definition: "{{ lookup('file', item ) | from_yaml }}"
+  loop:
+  - ./files/codeready_cr.yaml
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+      ssoSessionIdleTimeout: 28800
+      ssoSessionMaxLifespan: 28800
+    body_format: json
+    status_code: 204
+
+- name: create codeready users
+  include_tasks: add_che_user.yaml
+  vars:
+    user: "{{ item }}"
+  with_list: "{{ users }}"
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-global-operators.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-jupyterhub.yaml [hunk 14]
```diff
@@ -0,0 +1,14 @@
+--- 
+- name: Create jupyterhub instance
+  k8s:
+    state: present    
+    definition: "{{ lookup('file', './files/jupyterhub_kfdef.yaml' ) | from_yaml }}"
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-jupyterhub.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-mon.yaml [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+- name: Create operator subscription for Prometheus
+  k8s:
+    state: present
+    merge_type:
+    - strategic-merge
+    - merge
+    definition: "{{ lookup('file', item ) | from_yaml }}"
+  loop:
+  - ./files/prom_operatorgroup.yaml
+  - ./files/prom_subscription.yaml
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-mon.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-mon.yaml [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+- name: Create Prometheus CR
+  k8s:
+    state: present
+    merge_type:
+    - strategic-merge
+    - merge
+    definition: "{{ lookup('file', item ) | from_yaml }}"
+  loop:
+    - ./files/prom_cr.yaml
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-mon.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/dup_aws_credentials.yaml [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+      kind: "{{kind}}"
+      apiVersion: v1
+      metadata:
+        namespace: "{% if item[1]!='labs-infra' %}{{item[0].metadata.namespace.split('-') | first}}-{{item[1]}}{% else %}{{item[1]}}{% endif %}"
+        name: "{% if item[1]!='labs-infra' %}mlflow-obc{% else %}{{item[0].metadata.namespace.split('-') | first}}-{{item[0].metadata.name}}{% endif %}"
+      data: "{{item[0].data}}"
+  when: item[0].metadata.name == 'mlflow-obc' and item[0].metadata.namespace.endswith('-dev')
+  with_nested: 
+    - "{{results.resources}}"
+    - "{{namespaces}}"
```
[the remark this answers is not in the document]
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/dup_aws_credentials.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml (hunk 1 above)
Comment (eanylin) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml:133]: Updated

File: ansible/roles/ocp4-workload-mlops/templates/stage-pipeline.yaml.j2 [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+  resources:
+    inputs:
+    - name: source
+      type: git
+  workspaces:
+    - name: workspace
+      mountPath: /workspace
+  steps:
+  - name: setup-workspace
+    image: registry.redhat.io/openshift-pipelines-tech-preview/pipelines-git-init-rhel8@sha256:bb744988e87cc556873098885ee1a7fc1ef7b6c1ea24e3a4d7a5e518edb22838
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/templates/stage-pipeline.yaml.j2]: Updated.

File: ansible/roles/ocp4-workload-mlops/tasks/add-gog-repo.yaml [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    repo: "{{ repository.base }}/{{ repository.name }}"
+    dest: "{{ tmp_dir }}/{{ repository.name }}"
+    force: true
+
+- name: Push {{ repository.name }} to git repository in Gogs for {{ user }}
+  shell: |
+    git -c http.sslVerify=false  remote add {{ user }} https://{{ user }}:{{ gogs_pwd }}@gogs-labs-infra.{{ route_subdomain }}/{{ user }}/{{ repository.name }}.git
+    git -c http.sslVerify=false  config --local user.email "{{ user }}@example.com"
+    git -c http.sslVerify=false  config --local user.name "{{ user }}"
+    git -c http.sslVerify=false  push -f --all {{ user }}
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/add-gog-repo.yaml]: Fixed

File: ansible/roles/ocp4-workload-mlops/defaults/main.yml [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+# These binaries should exists on the ansible node
+binaries:
+  - oc
+  - curl
+  - git
+
+images_to_seed:
+  - name: nbviewer
+    uri: quay.io/odh-jupyterhub/nbviewer@sha256:21ce1db27e96305718a06359c86139680b8138966c614b4c94796a788eeac98c
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/defaults/main.yml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+- name: output workshop info guides
+  debug:
+    msg:
+    - "user.info: "
+    - "user.info: URL for attendees to get a username assigned to them and links to labs:"
+    - "user.info: http://get-a-username-labs-infra.{{ route_subdomain }}"
+    - "user.info: You should share this URL (or a shortlink for it) -- It is all they will need to get started!"
+    - "user.info: "
+    - "user.info: get-a-username's admin page is https://get-a-username-labs-infra.{{ route_subdomain }}/admin"
+    - "user.info: Admin login with admin/{{workshop_openshift_user_password}}"
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-dm.yaml [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+  - name: Create operator subscription for Decision Manager
+    k8s:
+      state: present
+      merge_type:
+      - strategic-merge
+      - merge
+      definition: "{{ lookup('file', item ) | from_yaml }}"
+    loop:
+    - ./files/dm_operatorgroup.yaml
+    - ./files/dm_subscription.yaml
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-dm.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+
+- name: wait 2 minutes and let the image download and be registered so workspaces start up
+  pause:
+      minutes: 2
+
+- name: Attempt to warm workspaces which failed to start
+  include_tasks: verify_che_workspace.yaml
+  vars:
+    user: "{{ item }}"
+  with_list: "{{ users }}"
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Updated.

File: ansible/roles/ocp4-workload-mlops/tasks/install-username-distribution.yaml [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+  shell: >
+    oc process -f -
+    -p DATABASE_SERVICE_NAME=redis
+    -p REDIS_PASSWORD=redis
+    -p VOLUME_CAPACITY=1Gi
+    -p MEMORY_LIMIT=1Gi
+    -p REDIS_VERSION=5
+    | oc create -n labs-infra  -f -
+  args:
+    stdin: "{{ lookup('file', './files/redis-template.yaml') }}"
```
[the remark this answers is not in the document]
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-username-distribution.yaml]: Updated.

File: ansible/roles/ocp4-workload-mlops/defaults/main.yml [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+become_override: no
+ocp_username: opentlc-mgr
+silent: False
+
+num_users: 10
+
+module_titles:
+  - { name: m1, title: "Click here to get started", path: /workshop/mlops/lab/kubernetes-overview }
+
+guid: 1002
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-mlops/defaults/main.yml:11]: There should not be a default value for `guid`. This should simply be required.

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml (hunk 1 above)
Comment (jkupferer) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml:133]: I still see `with_list` here?

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+
+- name: Pre-create and warm user workspaces
+  include_tasks: create_che_workspace.yaml
+  vars:
+    user: "{{ item }}"
+  with_list: "{{ users }}"
+
+- name: wait 2 minutes and let the image download and be registered so workspaces start up
+  pause:
+      minutes: 2
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml:137]: inconsistent indent for `minutes: 2`

File: ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml [hunk 27]
```diff
@@ -0,0 +1,43 @@
+---
+# Implement your Post Workload deployment tasks here
+- name: output workshop info
+  debug:
+    msg:
+    - "agnosticd_user_info: "
+    - "agnosticd_user_info: Provisioned for {{ num_users }} user(s)"
```
Comment (jkupferer) [inline ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml:7]: Usage of `agnosticd_user_info` is as a module to replace `debug`. This should be:

```
- name: output workshop info
  agnosticd_user_info:
    msg:
    - ""
    - "Provisioned for {{ num_users }} user(s)"
```

**Review by jkupferer — CHANGES_REQUESTED** [review]

File: ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml (hunk 27 above)
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/post_workload.yml:7]: Updated

File: ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml (hunk 26 above)
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/tasks/install-codeready.yaml]: Updated

File: ansible/roles/ocp4-workload-mlops/defaults/main.yml (hunk 25 above)
Comment (tsailiming) [reply ansible/roles/ocp4-workload-mlops/defaults/main.yml]: Updated

**Review by jkupferer — APPROVED** [review]

**Comment by tsailiming** [comment]
Merged latest from upstream/development and has tested on Dev env.

**Pull request merged**